File: core/memory/long_term.py
"""Long-Term Memory — 持久化记忆 + 时间衰减 + 情感索引 + 冲突检测。

从现有 episodic_memory.py 重构:
- 保留 EpisodicMemory dataclass (情感签名 + 显著性 + related_ids)
- 保留 hybrid search 接口框架
- 新增实际 embedding 函数注入
- 新增冲突检测 + 矛盾扫描 (参考 anda-hippocampus)
"""
from __future__ import annotations

import math
import sqlite3
import time

from .store import MemoryStore, MemoryRecord, ConsolidationReport


class LongTermMemory(MemoryStore):
    """长期记忆 — SQLite + 时间衰减 + 情感签名索引。

    容量 ~500 条。30 天半衰期。
    """

    def __init__(self, db_path: str = ":memory:", max_items: int = 500,
                 half_life_days: float = 30.0):
        self.db_path = db_path
        self.max_items = max_items
        self.half_life_days = half_life_days
        self._conn: sqlite3.Connection | None = None
        self._embedding_fn = None

    async def initialize(self) -> None:
        self._conn = sqlite3.connect(self.db_path)
        self._conn.execute("PRAGMA journal_mode=WAL")
        self._conn.execute("""
            CREATE TABLE IF NOT EXISTS ltm (
                record_id TEXT PRIMARY KEY,
                content TEXT NOT NULL,
                emotion TEXT DEFAULT '{}',
                significance REAL DEFAULT 0.5,
                event_type TEXT DEFAULT 'unknown',
                tags TEXT DEFAULT '[]',
                timestamp REAL,
                recall_count INTEGER DEFAULT 0,
                related_ids TEXT DEFAULT '[]',
                embedding BLOB,
                superseded INTEGER DEFAULT 0
            )
        """)
        self._conn.execute("""
            CREATE INDEX IF NOT EXISTS idx_ltm_emotion ON ltm(event_type)
        """)
        self._conn.execute("""
            CREATE INDEX IF NOT EXISTS idx_ltm_significance ON ltm(significance)
        """)
        self._conn.commit()

    def set_embedding_fn(self, fn):
        self._embedding_fn = fn

    async def store(self, record: MemoryRecord) -> str:
        import json

        rid = record.record_id or f"ltm_{int(time.time()*1000)}_{len(self)}"
        record.record_id = rid

        emb_blob = None
        if self._embedding_fn:
            try:
                emb = self._embedding_fn(record.content)
                import struct
                emb_blob = struct.pack(f'{len(emb)}f', *emb)
            except Exception:
                pass

        related = json.dumps(record.metadata.get("related_ids", []), ensure_ascii=False)

        self._conn.execute(
            "INSERT OR REPLACE INTO ltm VALUES (?,?,?,?,?,?,?,?,?,?,?)",
            (rid, record.content, json.dumps(record.emotional_signature, ensure_ascii=False),
             record.significance, record.event_type, json.dumps(record.tags, ensure_ascii=False),
             record.timestamp, record.recall_count, related, emb_blob, 0),
        )
        self._conn.commit()
        self._trim()
        return rid

    async def recall(self, query: str, n: int = 5) -> list[MemoryRecord]:
        """混合检索: 语义(0.3) + 时间衰减(0.3) + 显著性(0.4) + 类型加分。"""
        all_rows = self._conn.execute(
            "SELECT * FROM ltm WHERE superseded = 0 ORDER BY timestamp DESC LIMIT ?",
            (n * 5,),
        ).fetchall()

        now = time.time()
        scored: list[tuple[float, MemoryRecord]] = []

        for row in all_rows:
            r = self._row_to_record(row)
            sig_weight = r.significance * 0.4
            time_weight = self._time_decay(r, now) * 0.3

            # 语义相似度: 简单关键词匹配
            sem_weight = 0.0
            query_lower = query.lower()
            if query_lower in r.content.lower():
                sem_weight = 0.3
            else:
                keywords = query_lower.split()
                hits = sum(1 for kw in keywords if kw in r.content.lower())
                sem_weight = min(0.3, hits * 0.1)

            score = sem_weight + time_weight + sig_weight
            scored.append((score, r))

        scored.sort(key=lambda x: x[0], reverse=True)
        for _, r in scored[:n]:
            self._conn.execute(
                "UPDATE ltm SET recall_count = recall_count + 1 WHERE record_id = ?",
                (r.record_id,),
            )
        self._conn.commit()
        return [r for _, r in scored[:n]]

    async def search(self, embedding=None, filters=None, n=5) -> list[MemoryRecord]:
        if filters and filters.get("query"):
            return await self.recall(filters["query"], n)
        return await self.recall("", n)

    async def consolidate(self) -> ConsolidationReport:
        """长期记忆巩固: 去重 + 关联检测。"""
        report = ConsolidationReport()
        # 检测相邻相似记忆（简易去重）
        rows = self._conn.execute(
            "SELECT record_id, content FROM ltm WHERE superseded = 0 "
            "ORDER BY timestamp DESC LIMIT 50"
        ).fetchall()
        seen: set[str] = set()
        for rid, content in rows:
            # 简化: 完全相同内容标记为重复
            if content in seen:
                self._conn.execute(
                    "UPDATE ltm SET superseded = 1 WHERE record_id = ?", (rid,)
                )
                report.merged += 1
            seen.add(content)
        self._conn.commit()
        return report

    async def forget(self) -> int:
        """时间衰减淘汰: 30 天以上的低显著性记忆。"""
        cutoff = time.time() - self.half_life_days * 86400
        cursor = self._conn.execute(
            "DELETE FROM ltm WHERE timestamp < ? AND significance < 0.3",
            (cutoff,),
        )
        self._conn.commit()
        return cursor.rowcount

    def detect_contradictions(self) -> list[dict]:
        """矛盾检测 (参考 anda): 同实体、相反情感 → 标记冲突。"""
        import json
        conflicts = []
        rows = self._conn.execute(
            "SELECT * FROM ltm WHERE superseded = 0 ORDER BY timestamp DESC LIMIT 100"
        ).fetchall()

        for i in range(len(rows)):
            for j in range(i + 1, len(rows)):
                r1 = self._row_to_record(rows[i])
                r2 = self._row_to_record(rows[j])
                # 检查情感对立 (positive vs negative)
                e1 = r1.emotional_signature
                e2 = r2.emotional_signature
                if (e1.get("joy", 0) > 0.5 and e2.get("sadness", 0) > 0.5) or \
                   (e1.get("trust", 0) > 0.5 and e2.get("fear", 0) > 0.5):
                    conflicts.append({
                        "record_a": r1.record_id,
                        "record_b": r2.record_id,
                        "type": "emotional_contradiction",
                    })
        return conflicts

    def promote_candidates(self) -> list[MemoryRecord]:
        """recall>=5 + sig>0.8 → 候选提升到 Core 图谱。"""
        rows = self._conn.execute(
            "SELECT * FROM ltm WHERE recall_count >= 5 AND significance >= 0.8 AND superseded = 0"
        ).fetchall()
        return [self._row_to_record(r) for r in rows]

    def _time_decay(self, record: MemoryRecord, now: float) -> float:
        age_days = (now - record.timestamp) / 86400.0
        return math.exp(-age_days / self.half_life_days)

    def _trim(self):
        count = self._conn.execute("SELECT COUNT(*) FROM ltm").fetchone()[0]
        if count > self.max_items:
            excess = count - self.max_items
            self._conn.execute(
                "DELETE FROM ltm WHERE rowid IN ("
                "SELECT rowid FROM ltm ORDER BY significance ASC, timestamp ASC LIMIT ?)",
                (excess,),
            )
            self._conn.commit()

    def _row_to_record(self, row) -> MemoryRecord:
        import json
        return MemoryRecord(
            record_id=row[0], content=row[1],
            emotional_signature=json.loads(row[2]) if row[2] else {},
            significance=row[3], event_type=row[4],
            tags=json.loads(row[5]) if row[5] else [],
            timestamp=row[6], recall_count=row[7],
            metadata={
                "related_ids": json.loads(row[8]) if row[8] else [],
                "embedding": row[9],
                "superseded": bool(row[10]),
            },
        )

    async def on_session_end(self) -> None:
        if self._conn:
            self._conn.commit()

    async def shutdown(self) -> None:
        if self._conn:
            self._conn.commit()
            self._conn.close()
            self._conn = None

    def __len__(self) -> int:
        if self._conn:
            return self._conn.execute("SELECT COUNT(*) FROM ltm WHERE superseded = 0").fetchone()[0]
        return 0

File: src/character/memory/long-term.ts
/** Long-Term Memory — SQLite + time decay. better-sqlite3 for Node.js */
import Database from "better-sqlite3";
import { MemoryStore, MemoryRecord, createMemoryRecord, ConsolidationReport, createConsolidationReport } from "./store";

export class LongTermMemory extends MemoryStore {
  private dbPath: string; private maxItems: number; private halfLifeDays: number;
  private _db: Database | null = null;

  constructor(dbPath = ":memory:", maxItems = 500, halfLifeDays = 30) {
    super(); this.dbPath = dbPath; this.maxItems = maxItems; this.halfLifeDays = halfLifeDays;
  }

  get length(): number {
    return this._db ? (this._db.prepare("SELECT COUNT(*) as c FROM ltm WHERE superseded=0").get() as any).c : 0;
  }

  async initialize(): Promise<void> {
    this._db = new Database(this.dbPath);
    this._db.pragma("journal_mode = WAL");
    this._db.exec(`CREATE TABLE IF NOT EXISTS ltm (
      record_id TEXT PRIMARY KEY, content TEXT NOT NULL, emotion TEXT DEFAULT '{}',
      significance REAL DEFAULT 0.5, event_type TEXT DEFAULT 'unknown', tags TEXT DEFAULT '[]',
      timestamp REAL, recall_count INTEGER DEFAULT 0, related_ids TEXT DEFAULT '[]',
      memory_type TEXT DEFAULT 'episodic', confidence REAL DEFAULT 0.7,
      superseded INTEGER DEFAULT 0, superseded_by TEXT, embedding BLOB
    )`);
    this._db.exec("CREATE INDEX IF NOT EXISTS idx_ltm_emotion ON ltm(event_type)");
    this._db.exec("CREATE INDEX IF NOT EXISTS idx_ltm_sig ON ltm(significance)");
    this._db.exec("CREATE INDEX IF NOT EXISTS idx_ltm_conf ON ltm(confidence)");
  }

  async store(record: MemoryRecord): Promise<string> {
    const rid = record.recordId || `ltm_${Date.now()}_${this.length}`;
    record.recordId = rid;
    this._db!.prepare("INSERT OR REPLACE INTO ltm VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?)").run(
      rid, record.content, JSON.stringify(record.emotionalSignature), record.significance,
      record.eventType, JSON.stringify(record.tags), record.timestamp, record.recallCount,
      JSON.stringify(record.metadata?.relatedIds ?? []), record.memoryType,
      record.confidence, record.superseded ? 1 : 0, record.supersededBy, null,
    );
    this._trim(); return rid;
  }

  async recall(query: string, n = 5): Promise<MemoryRecord[]> {
    const rows = this._db!.prepare(
      "SELECT * FROM ltm WHERE superseded=0 ORDER BY (significance * confidence) DESC LIMIT ?"
    ).all(n * 5) as any[];
    const now = Date.now() / 1000;
    const scored: Array<[number, MemoryRecord]> = [];
    for (const row of rows) {
      const r = this._rowToRecord(row);
      const ql = query.toLowerCase(); let sem = 0;
      if (r.content.toLowerCase().includes(ql)) sem = 0.3;
      else { const hits = ql.split(/\s+/).filter(kw => r.content.toLowerCase().includes(kw)).length; sem = Math.min(0.3, hits * 0.1); }
      scored.push([r.significance * 0.4 + this._timeDecay(r, now) * 0.3 + sem, r]);
    }
    scored.sort((a, b) => b[0] - a[0]);
    for (const [_, r] of scored.slice(0, n)) this._db!.prepare("UPDATE ltm SET recall_count=recall_count+1 WHERE record_id=?").run(r.recordId);
    return scored.slice(0, n).map(([_, r]) => r);
  }

  async search(_e?: number[] | null, filters?: Record<string, unknown> | null, n = 5): Promise<MemoryRecord[]> {
    return filters?.query ? this.recall(filters.query as string, n) : this.recall("", n);
  }

  async consolidate(): Promise<ConsolidationReport> {
    const report = createConsolidationReport();
    const rows = this._db!.prepare("SELECT record_id, content FROM ltm WHERE superseded=0 ORDER BY timestamp DESC LIMIT 50").all() as any[];
    const seen = new Set<string>();
    for (const row of rows) {
      const rid: string = row[0] ?? row.record_id;
      const content: string = row[1] ?? row.content;
      if (seen.has(content)) {
        this._db!.prepare("UPDATE ltm SET superseded=1 WHERE record_id=?").run(rid);
        report.merged++;
      }
      seen.add(content);
    }
    return report;
  }

  async forget(): Promise<number> {
    return this._db!.prepare("DELETE FROM ltm WHERE timestamp < ? AND significance < 0.3").run(Date.now() / 1000 - this.halfLifeDays * 86400).changes;
  }

  detectContradictions(): Array<Record<string, string>> {
    const rows = this._db!.prepare("SELECT * FROM ltm WHERE superseded=0 ORDER BY timestamp DESC LIMIT 100").all() as any[];
    const conflicts: Array<Record<string, string>> = [];
    for (let i = 0; i < rows.length; i++) {
      for (let j = i + 1; j < rows.length; j++) {
        const e1 = JSON.parse(rows[i][2] || "{}"), e2 = JSON.parse(rows[j][2] || "{}");
        if ((e1.joy > 0.5 && e2.sadness > 0.5) || (e1.trust > 0.5 && e2.fear > 0.5))
          conflicts.push({ recordA: rows[i][0], recordB: rows[j][0], type: "emotional_contradiction" });
      }
    }
    return conflicts;
  }

  promoteCandidates(): MemoryRecord[] {
    return (this._db!.prepare("SELECT * FROM ltm WHERE recall_count>=5 AND significance>=0.8 AND superseded=0").all() as any[]).map((r: any) => this._rowToRecord(r));
  }

  private _timeDecay(r: MemoryRecord, now: number): number { return Math.exp(-(now - r.timestamp) / (this.halfLifeDays * 86400)); }
  private _trim(): void {
    const c = (this._db!.prepare("SELECT COUNT(*) as c FROM ltm").get() as any).c;
    if (c > this.maxItems) this._db!.prepare("DELETE FROM ltm WHERE rowid IN (SELECT rowid FROM ltm ORDER BY significance ASC, timestamp ASC LIMIT ?)").run(c - this.maxItems);
  }
  /** Update confidence — reinforced by recall or contradicted by new evidence. */
  updateConfidence(recordId: string, delta: number): void {
    this._db!.prepare(
      "UPDATE ltm SET confidence = MAX(0.0, MIN(1.0, confidence + ?)) WHERE record_id=?"
    ).run(delta, recordId);
  }

  /** Mark a fact as superseded by a newer version. Never silently delete. */
  markSuperseded(recordId: string, byRecordId: string): void {
    this._db!.prepare(
      "UPDATE ltm SET superseded=1, superseded_by=?, confidence=0.1 WHERE record_id=?"
    ).run(byRecordId, recordId);
  }

  /** Decay confidence of old, unverified facts. */
  decayConfidence(halfLifeSeconds: number, now: number): number {
    const lambda = Math.LN2 / halfLifeSeconds;
    return this._db!.prepare(
      `UPDATE ltm SET confidence = confidence * ?
       WHERE superseded=0 AND recall_count=0 AND confidence > 0.2
       AND (? - timestamp) > ?`
    ).run(Math.exp(-lambda), now, halfLifeSeconds).changes;
  }

  /** Compress old records: full text → short summary + emotional vector. */
  compressOld(minAgeSeconds: number, now: number): number {
    return this._db!.prepare(
      `UPDATE ltm SET significance = significance * 0.5
       WHERE superseded=0 AND confidence < 0.3
       AND (? - timestamp) > ? AND recall_count = 0`
    ).run(now, minAgeSeconds).changes;
  }

  private _rowToRecord(row: any): MemoryRecord {
    return createMemoryRecord({
      recordId: row[0], content: row[1],
      emotionalSignature: JSON.parse(row[2] || "{}"),
      significance: row[3], eventType: row[4],
      tags: JSON.parse(row[5] || "[]"), timestamp: row[6],
      recallCount: row[7],
      memoryType: (row[9] as any) ?? "episodic",
      confidence: row[10] ?? 0.7,
      superseded: !!row[11],
      supersededBy: row[12] ?? null,
      metadata: { relatedIds: JSON.parse(row[8] || "[]"), embedding: row[13] },
    });
  }
}
